pkg/scheduler/plugins/elasticquota/plugin.go

```diff
@@ -0,0 +1,565 @@
+func (s *SnapshotState) Clone() framework.StateData {
+	return &SnapshotState{
+		elasticQuotaInfos: s.elasticQuotaInfos.clone(),
+	}
+}
+
+var (
+	_ framework.PreFilterPlugin  = &Plugin{}
+	_ framework.PostFilterPlugin = &Plugin{}
+	_ framework.ReservePlugin    = &Plugin{}
+)
+
+func New(args runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	pluginArgs, ok := args.(*config.ElasticQuotaArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type GangSchedulingArgs, got %T", args)
+	}
+	if err := validation.ValidateElasticQuotaArgs(pluginArgs); err != nil {
+		return nil, err
+	}
+
+	client, ok := handle.(versioned.Interface)
+	if !ok {
+		client = versioned.NewForConfigOrDie(handle.KubeConfig())
+	}
+	scheSharedInformerFactory := externalversions.NewSharedInformerFactory(client, 0)
+	elasticQuotaInformer := scheSharedInformerFactory.Scheduling().V1alpha1().ElasticQuotas()
+
+	elasticQuota := &Plugin{
+		handle:            handle,
+		pluginArgs:        pluginArgs,
+		podLister:         handle.SharedInformerFactory().Core().V1().Pods().Lister(),
+		quotaLister:       elasticQuotaInformer.Lister(),
+		pdbLister:         getPDBLister(handle.SharedInformerFactory()),
+		nodeLister:        handle.SharedInformerFactory().Core().V1().Nodes().Lister(),
+		groupQuotaManager: core.NewGroupQuotaManager(pluginArgs.SystemQuotaGroupMax, pluginArgs.DefaultQuotaGroupMax),
+		nodeResourceMap:   make(map[string]struct{}),
+		elasticQuotaInfos: NewSimpleElasticQuotaInfos(),
+		failOver:          make(chan struct{}),
+	}
+	elasticQuota.groupQuotaManager.SetScaleMinQuotaEnabled(true)
+	elasticQuota.quotaOverUsedRevokeController = NewQuotaOverUsedRevokeController(
+		*elasticQuota.pluginArgs.ContinueOverUseCountTriggerEvict, elasticQuota.groupQuotaManager)
+
+	podInformer := handle.SharedInformerFactory().Core().V1().Pods()
+	podInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    elasticQuota.OnPodAdd,
+			UpdateFunc: elasticQuota.OnPodUpdate,
+			DeleteFunc: elasticQuota.OnPodDelete,
+		},
+	)
+
+	nodeInformer := handle.SharedInformerFactory().Core().V1().Nodes().Informer()
+	nodeInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    elasticQuota.OnNodeAdd,
+			UpdateFunc: elasticQuota.OnNodeUpdate,
+			DeleteFunc: elasticQuota.OnNodeDelete,
+		})
+
+	elasticQuotaInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    elasticQuota.OnQuotaAdd,
+			UpdateFunc: elasticQuota.OnQuotaUpdate,
+			DeleteFunc: elasticQuota.OnQuotaDelete,
+		})
+
+	elasticQuota.elasticQuotaController = NewElasticQuotaController(kubernetes.NewForConfigOrDie(handle.KubeConfig()),
+		elasticQuotaInformer.Lister(), client, elasticQuota.groupQuotaManager)
+	ctx := context.TODO()
+
+	// need to recover quota first
+	scheSharedInformerFactory.Start(ctx.Done())
+	scheSharedInformerFactory.WaitForCacheSync(ctx.Done())
+	close(elasticQuota.failOver)
```

Comment:
just start the scheSharedInformerFactory before registering Node/Pod EventHandler.  

If the other plugins have already started the PodInformer, when you register the Pod EventHandler, informer will push all stored Pods into the new EventHandler.  Then the handler waits for the channel `failover`, possible deadlock?

Reply:
fixed